source/common/router/config_impl.cc

```diff
@@ -1816,11 +1816,18 @@ RouteSpecificFilterConfigConstSharedPtr PerFilterConfigs::createRouteSpecificFil
     Server::Configuration::ServerFactoryContext& factory_context,
     ProtobufMessage::ValidationVisitor& validator) {
   bool is_optional = (optional_http_filters.find(name) != optional_http_filters.end());
-  auto factory = Envoy::Config::Utility::getAndCheckFactoryByName<
-      Server::Configuration::NamedHttpFilterConfigFactory>(name, is_optional);
+  auto factory =
+      Envoy::Config::Utility::getFactoryByType<Server::Configuration::NamedHttpFilterConfigFactory>(
+          typed_config);
   if (factory == nullptr) {
-    ENVOY_LOG(warn, "Can't find a registered implementation for http filter '{}'", name);
-    return nullptr;
+    if (is_optional) {
+      ENVOY_LOG(warn, "Can't find a registered implementation for http filter '{}'", name);
+      return nullptr;
+    } else {
+      throw EnvoyException(
+          fmt::format("Didn't find a registered implementation for '{}' with type URL: '{}'", name,
+                      Envoy::Config::Utility::getFactoryType(typed_config)));
+    }
```

Comment:
But guard is still ok. Because this PR did lead to some behavior changes. In the new implemention, we cannot use filter name to get the factory of route config.

Reply:
OK yeah I would still do the guard and clearly document the change. I regret not doing a guard and/or warning on the other PR that @kyessenov did to require type URL as I'm guessing this broke almost every user which is unfortunate.

Reply:
Get it. 😄